### DISCUSSION TITLE
RFC: New Names Proposal

### DISCUSSION THREAD

**Discussion by jwass** [OvertureMaps/data#328] [Discussion]
Seeking feedback on a new Overture names proposal. Please include any comments/feedback in this discussion.

After working with Overture data releases, one issue that's come up is friction working with the names schema. Names for a feature can be complex with different variants (common, official, alternates, etc.) and different translations. Furthermore, transportation segments need to specify a name that applies over only a section of the segment. 

This new names schema is designed to allow ease of access to a feature's default/primary name as well common names and their  translations. This design sets out to simplify SQL queries and data structures for the most common access patterns. The names structure also supports non-common and other concepts (abbreviations, official names, linear-referenced, ...) that require a more complex schema and associated code when using them.

Here's a fictional example that demonstrates the different names (Note this is not a realistic feature since a palace won't have linear-referenced names)
```
{
  "names": {
    "primary": "I used to be called local",
    "common": {
      "en": "Buckingham Palace"
    },
    "rules": [
      {
        "variant": "official",
        "value": "Her Majesty's Most Smashingly Groovy Palace",
        "language": "en-gb",
      },
      {
        "variant": "common",
        "value": "Pile of Bones Road",
        "language": "en-ca",
        "side": "left",
        "at": [0, 0.5]
      }
    ]
  }
}
```
Pull request here: https://github.com/OvertureMaps/schema/pull/98

#### Get the "default" name of a feature.
Old: `names.common[1].value`
New: `names.primary`

#### Find the Spanish name of a feature
Old (Presto) `FILTER(names.common, n -> n.language = 'es')[1].value`
Note this can fail if the Spanish name isn't present causing even more complex code to check.

New: `names.common['es']`

#### Use the English name and if not present, use the default name
This is actually pretty difficult in SQL with the old schema so I'll save the pain of typing it out.
New: `COALESCE(names.common['en'], names.primary)`

**Comment by jwass** [comment]
@mtravis I know you've mentioned difficulty of using the current names schema. Curious if you have any thoughts/feedback here.

**Reply by mtravis** [reply]
@jwass yes I had issues with names in admin as I was wrongly assuming every entry would have an `en` entry so some features weren't coming back in my queries. 

Having something like primary name baked in to Overture would be really useful.

**Reply by mtravis** [reply]
COALESCE(names.common['en'], names.primary) would be perfect.

**Comment by jwass** [comment]
Closing since https://github.com/OvertureMaps/schema/pull/98 was merged.
